File: v8-core/src/experts/fib_rsi_bb_confluence.rs
//! fib_rsi_bb_confluence: evaluate() port — mirror src/v8/experts/fib_rsi_bb_confluence.py
//! bit-for-bit (PARITY_AND_IDENTITY_SPEC §3; COMPUTE_CORE_SPEC §8 S4).
//! Three-leg confluence: BB fade zone (close in 2-SD..3-SD), Wilder-RSI dip
//! recovery (state feature + local series), fib 0.786 reclaim of the prior
//! impulse. Variant a (STRICT: all three legs agree) — the harness instantiates
//! `FibRsiBbConfluenceExpert()` with the default variant_id 'a'. The D-026
//! anchor is the consecutive run of bars where the SAME confluence predicate
//! holds, via base::find_setup_anchor.

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::HistBar;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["oscillator", "location", "volatility", "history"];

// Declared, frozen constants, inherited verbatim from the registered families
// (D-036/D-046 pattern; never fitted on the dev window).
const BB_BASE_N: usize = 20;  // bollinger_reversion.py:43
const RSI_OS: f64 = 30.0;     // rsi_stoch_reversion.py:45
const RSI_OB: f64 = 70.0;     // rsi_stoch_reversion.py:46
// The confluence fib leg uses the DEEPEST retracement (structural co-occurrence
// argument in the module docstring); it doubles as the post-entry
// deep-correction reference, as in fib_retracement_continuation.
const FIB_RATIO: f64 = 0.786;
// The harness runs FibRsiBbConfluenceExpert() with no variant argument, so the
// instance variant_id is the class default 'a' (STRICT).
const VARIANT: &str = "a";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Dir {
    Long,
    Short,
}

impl Dir {
    fn label(self) -> &'static str {
        match self {
            Dir::Long => "LONG",
            Dir::Short => "SHORT",
        }
    }
}

/// Python `_mean`: `sum(values) / len(values)`. Builtin `sum()` is plain
/// left-to-right double accumulation (CPython float fast path), so Rust's
/// `Iterator::sum` agrees bit-for-bit.
fn mean(values: &[f64]) -> f64 {
    values.iter().sum::<f64>() / values.len() as f64
}

/// Python `_std_pop`: `(sum((v - m) ** 2 for v in values) / len(values)) ** 0.5`.
/// `** 2` / `** 0.5` are libm pow — use `.powf`, not `x * x` / `.sqrt()`.
fn std_pop(values: &[f64]) -> f64 {
    let m = mean(values);
    (values.iter().map(|v| (*v - m).powf(2.0)).sum::<f64>() / values.len() as f64).powf(0.5)
}

/// Python `_bb_series`: per-bar (mid, sd) of the trailing 20 closes; None in
/// warmup. A degenerate window still yields a pair; `sd <= 0` is rejected by
/// the vote (mirrors Python, which also computes the pair and rejects there).
fn bb_series(hist: &[HistBar]) -> Vec<Option<(f64, f64)>> {
    let closes: Vec<f64> = hist.iter().map(|b| b.close).collect();
    let mut out = Vec::with_capacity(closes.len());
    for i in 0..closes.len() {
        if i >= BB_BASE_N - 1 {
            let win = &closes[i - BB_BASE_N + 1..i + 1];
            out.push(Some((mean(win), std_pop(win))));
        } else {
            out.push(None);
        }
    }
    out
}

/// Python `_rsi_value`.
fn rsi_value(avg_gain: f64, avg_loss: f64) -> f64 {
    if avg_loss == 0.0 {
        return if avg_gain > 0.0 { 100.0 } else { 50.0 };
    }
    if avg_gain == 0.0 {
        return 0.0;
    }
    100.0 - 100.0 / (1.0 + avg_gain / avg_loss)
}

/// Python `_rsi_per_bar` (Wilder RSI over the local window; None before the
/// seed). Identical formula to marketstate's rsi14 over the local window.
fn rsi_per_bar(closes: &[f64], period: usize) -> Vec<Option<f64>> {
    if closes.len() < period + 1 {
        return vec![None; closes.len()];
    }
    let deltas: Vec<f64> = closes.windows(2).map(|w| w[1] - w[0]).collect();
    let gains: Vec<f64> = deltas.iter().map(|d| d.max(0.0)).collect();
    let losses: Vec<f64> = deltas.iter().map(|d| (-d).max(0.0)).collect();
    let mut avg_gain = gains[..period].iter().sum::<f64>() / period as f64;
    let mut avg_loss = losses[..period].iter().sum::<f64>() / period as f64;
    let mut out: Vec<Option<f64>> = vec![None; period];
    out.push(Some(rsi_value(avg_gain, avg_loss)));
    for i in period..deltas.len() {
        avg_gain = (avg_gain * (period as f64 - 1.0) + gains[i]) / period as f64;
        avg_loss = (avg_loss * (period as f64 - 1.0) + losses[i]) / period as f64;
        out.push(Some(rsi_value(avg_gain, avg_loss)));
    }
    out
}

/// Python `_bb_vote_at` (bollinger_reversion Setup 2): close between 2-SD and
/// 3-SD of the 20-bar SMA. Degenerate band (sd <= 0) is no level; a close
/// beyond 3-SD is a trend, not a reversion.
fn bb_vote_at(bb: &[Option<(f64, f64)>], i: usize, bar: &HistBar) -> Option<Dir> {
    if i < BB_BASE_N - 1 {
        return None;
    }
    let (mid, sd) = match bb[i] {
        Some(p) => p,
        None => return None,
    };
    if sd <= 0.0 {
        return None;
    }
    let close = bar.close;
    if mid - 3.0 * sd < close && close <= mid - 2.0 * sd {
        return Some(Dir::Long);
    }
    if mid + 2.0 * sd <= close && close < mid + 3.0 * sd {
        return Some(Dir::Short);
    }
    None
}

/// Python `_rsi_vote_at` (rsi_stoch_reversion variant a): the newest run sits
/// on the reverted side of its extreme AND the run start's predecessor is on
/// the extreme side. LONG checked first, mirroring the source's return-early
/// order.
fn rsi_vote_at(rsi: &[Option<f64>], i: usize) -> Option<Dir> {
    if i >= rsi.len() {
        return None;
    }
    let cur = match rsi[i] {
        Some(v) => v,
        None => return None,
    };
    if cur > RSI_OS {
        let mut s = i;
        while s > 0 && rsi[s - 1].map(|v| v > RSI_OS).unwrap_or(false) {
            s -= 1;
        }
        if s > 0 && rsi[s - 1].map(|v| v <= RSI_OS).unwrap_or(false) {
            return Some(Dir::Long);
        }
    }
    if cur < RSI_OB {
        let mut s = i;
        while s > 0 && rsi[s - 1].map(|v| v < RSI_OB).unwrap_or(false) {
            s -= 1;
        }
        if s > 0 && rsi[s - 1].map(|v| v >= RSI_OB).unwrap_or(false) {
            return Some(Dir::Short);
        }
    }
    None
}

/// Python `_fib_vote_at`: a close that reclaimed the deepest retracement of
/// the intact impulse. `fib_direction`/`fib_level` are frozen from the state's
/// fib_levels at detection, so the anchor scan uses the same level.
fn fib_vote_at(fib_direction: f64, fib_level: f64, bar: &HistBar) -> Option<Dir> {
    if fib_direction == 1.0 {
        if bar.close > fib_level && bar.low <= fib_level {
            return Some(Dir::Long);
        }
        None
    } else {
        if bar.close < fib_level && bar.high >= fib_level {
            return Some(Dir::Short);
        }
        None
    }
}

/// Python `_confluence_vote`. Variant a (STRICT): all three legs vote the same
/// direction. Variant b (MAJORITY): at least two of the three agree.
fn confluence_vote(variant: &str, votes: &[Option<Dir>; 3]) -> Option<Dir> {
    if variant == "a" {
        if let (Some(a), Some(b), Some(c)) = (votes[0], votes[1], votes[2]) {
            if a == b && b == c {
                return Some(a);
            }
        }
        return None;
    }
    let longs = votes.iter().filter(|v| **v == Some(Dir::Long)).count();
    let shorts = votes.iter().filter(|v| **v == Some(Dir::Short)).count();
    if longs >= 2 {
        return Some(Dir::Long);
    }
    if shorts >= 2 {
        return Some(Dir::Short);
    }
    None
}

/// Python `_retracement_level`: the level for `ratio` from the self-describing
/// fib tuple `(anchor, direction, retr, ext)`; None when absent.
fn retracement_level(fibs: &[serde_json::Value], ratio: f64) -> Option<f64> {
    let retr = match fibs.get(2).and_then(|v| v.as_array()) {
        Some(r) => r,
        None => return None,
    };
    for pair in retr {
        let r = match pair.get(0).and_then(|v| v.as_f64()) {
            Some(v) => v,
            None => continue,
        };
        if (r - ratio).abs() < 1e-9 {
            return pair.get(1).and_then(|v| v.as_f64());
        }
    }
    None
}

pub fn fib_rsi_bb_confluence(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    let sym = fm.symbol;
    // Python `_need`: every required feature key present, else NO_HABITAT.
    let need = ["close", "atr", "history", "bb_mid", "bb_upper", "bb_lower",
                "rsi14", "fib_levels"];
    for k in need {
        if !fm.features.contains_key(k) {
            return no_habitat(expert_id, version, fm.as_of);
        }
    }
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    // `atr is None` -> NO_HABITAT (Python checks the value, not just the key).
    let atr = match fm.value("atr") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    // `not hist_value` -> NO_HABITAT (fm.history mirrors Python's self._hist).
    if fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }
    // `not isinstance(fibs, tuple) or len(fibs) != 4` -> NO_HABITAT.
    let fibs = match fm.features.get("fib_levels") {
        Some(f) => &f.value,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let fib_arr = match fibs.as_array() {
        Some(a) if a.len() == 4 => a,
        _ => return no_habitat(expert_id, version, fm.as_of),
    };
    // Fib warmup needs a confirmed swing pair (n_close >= 21 in marketstate);
    // a shorter window cannot host the confluence (habitat-unavailable).
    if fm.history.len() < 21 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let anchor_price = match fib_arr[0].as_f64() {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let direction = match fib_arr[1].as_f64() {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if !(direction == 1.0 || direction == -1.0) || anchor_price <= 0.0 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let fib_level = match retracement_level(fib_arr, FIB_RATIO) {
        Some(l) => l,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let fib_direction = direction;

    let bb = bb_series(&fm.history);
    let closes: Vec<f64> = fm.history.iter().map(|b| b.close).collect();
    let rsi = rsi_per_bar(&closes, 14);

    // Current-bar votes. The RSI leg additionally requires the STATE feature
    // to clear the threshold (the full-series Wilder value can disagree with
    // the local window near the boundary; the conservative gate from
    // rsi_stoch_reversion).
    let last = fm.history.len() - 1;
    let bb_v = bb_vote_at(&bb, last, &fm.history[last]);
    let mut rsi_v = rsi_vote_at(&rsi, last);
    let rsi_feat = match fm.value("rsi14") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if rsi_v == Some(Dir::Long) && !(rsi_feat > RSI_OS) {
        rsi_v = None;
    }
    if rsi_v == Some(Dir::Short) && !(rsi_feat < RSI_OB) {
        rsi_v = None;
    }
    let fib_v = fib_vote_at(fib_direction, fib_level, &fm.history[last]);
    let direction_sig = match confluence_vote(VARIANT, &[bb_v, rsi_v, fib_v]) {
        Some(d) => d,
        None => return no_setup(expert_id, version, fm.as_of),
    };

    // D-026 anchor: the consecutive run of bars where the same confluence
    // predicate holds (base.py find_setup_anchor, newest-false + 1).
    let confluence_at = |i: usize, b: &HistBar| {
        let votes = [
            bb_vote_at(&bb, i, b),
            rsi_vote_at(&rsi, i),
            fib_vote_at(fib_direction, fib_level, b),
        ];
        confluence_vote(VARIANT, &votes).is_some()
    };
    let anchor = find_setup_anchor(&fm.history, &confluence_at);

    // Geometry, frozen at detection (D-042 prior_*_ref pattern). The 78.6%
    // level doubles as the pre-entry invalidation and the post-entry
    // deep-correction reference; the frozen 3-SD band is the reversion-premise
    // reference. `variant` separates the episode keys.
    let mid = match fm.value("bb_mid") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let upper = match fm.value("bb_upper") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let lower = match fm.value("bb_lower") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let mut geometry = geom(vec![
        ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
        ("target_r", serde_json::json!(1.0)),
        ("stop_r", serde_json::json!(1.0)),
        ("expiry_bars", serde_json::json!(8)),
        ("atr_ref", serde_json::json!(atr)),
        ("variant", serde_json::json!(VARIANT)),
    ]);
    match direction_sig {
        Dir::Long => {
            geometry.insert("prior_low_ref".to_string(), serde_json::json!(fib_level));
            geometry.insert("lower_3sd_ref".to_string(),
                            serde_json::json!(mid - 1.5 * (mid - lower)));
        }
        Dir::Short => {
            geometry.insert("prior_high_ref".to_string(), serde_json::json!(fib_level));
            geometry.insert("upper_3sd_ref".to_string(),
                            serde_json::json!(mid + 1.5 * (upper - mid)));
        }
    }
    let fingerprint = format!("{sym}:{}:{}:{:.6}:{:.6}",
                              VARIANT, direction_sig.label(), fib_level, close);
    let draft = Draft {
        direction: direction_sig.label().to_string(),
        birth_time: fm.as_of,
        risk_geometry: geometry,
    };
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}

File: src/v8/experts/fib_rsi_bb_confluence.py
"""Fibonacci + RSI + Bollinger confluence behavior family
(`fib_rsi_bb_confluence`).

Hypothesis (mechanism `confluence_reversion_continuation`): when THREE
independent indicator families point the same way at the same bar, the setup
carries more conviction than any single family alone. The three legs are the
registered families' own idioms, verbatim:

- **Bollinger leg** (`volatility_band` / `band_reversion`, Ch12 p481-482):
  a close in the 2-SD..3-SD fade zone — price stretched below the 20-SMA by
  two sigma is a reversion candidate, not a breakdown (a close beyond 3-SD is
  a trend).
- **RSI leg** (`oscillator_reversion` / `overbought_oversold_reversion`,
  Ch8.6 p259-262): the oscillator dipped below oversold (30) and recovered
  above it — the first sign the extreme is reverting. The feature AND the
  local Wilder series must both clear the threshold (the conservative gate
  from `rsi_stoch_reversion`; Wilder RSI over a short window can disagree
  with the full-series feature near the boundary).
- **Fibonacci leg** (`fib_level_reaction` / `fib_retracement_continuation`,
  Ch10.4-10.7): a close that reclaimed a retracement level of the prior
  impulse. The level is the DEEPEST retracement (78.6%) — the deepest
  support of an up-impulse (mirror for shorts). This is a structural choice,
  not a fit: the fade-zone close (below mid - 2*sigma) can only co-occur
  with a retracement level that sits below the 20-SMA's lower band; of the
  standard ratios only 0.786 does so, so a shallow 0.382 reclaim is
  geometrically near-impossible as a three-leg confluence (the co-occurrence
  was computed before the experiment ran, and the level is frozen).

Variants (all frozen; D-044 lists every implemented variant):
  a  STRICT: all three legs vote the same direction, or no signal.
  b  MAJORITY: at least two of the three legs agree, or no signal.

The STRICT variant is the reason the family exists: it answers "does
requiring ALL THREE families to agree beat requiring only TWO?" (the
confluence vs. its own relaxation). Variant b is the same family's
relaxation — one multiplicity unit, one declared search universe.

Risk geometry: the family default 1R:1R:8bar with `atr_ref` (D-028), frozen
at detection. The 78.6% level becomes `prior_low_ref`/`prior_high_ref` (the
D-042 `prior_*_ref` pattern — the pre-entry invalidation AND the post-entry
deep-correction reference), and the frozen 3-SD band becomes
`lower_3sd_ref`/`upper_3sd_ref` (the reversion-premise reference: a close
beyond the 3-SD band is a trend, not a reversion). The `variant` key in the
geometry separates the two variants' episode keys (`lab._geometry_version`
excludes only `atr_ref`/`prior_*_ref`; without it variant-b candidates are
suppressed as duplicates of variant-a).
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert

# Declared, frozen constants, inherited verbatim from the registered
# families (D-036 pattern; never fitted on the dev window).
BB_BASE_N = 20                       # bollinger_reversion.py:43
RSI_OS = 30.0                        # rsi_stoch_reversion.py:45
RSI_OB = 70.0                        # rsi_stoch_reversion.py:46
# The confluence fib leg uses the DEEPEST retracement (the structural
# co-occurrence argument in the module docstring); it doubles as the
# post-entry deep-correction reference, as in fib_retracement_continuation.
FIB_RATIO = 0.786
DEEP_RETRACEMENT = 0.786


def _mean(values):
    return sum(values) / len(values)


def _std_pop(values):
    m = _mean(values)
    return (sum((v - m) ** 2 for v in values) / len(values)) ** 0.5


def _bb_series(hist):
    """Per-bar (mid, sd) of the trailing 20 closes; None in warmup. Copy of
    bollinger_reversion._bb_series (private there; duplicated with
    attribution so the confluence recomputes the same series locally)."""
    closes = [b[4] for b in hist]
    out = []
    for i in range(len(closes)):
        if i >= BB_BASE_N - 1:
            win = closes[i - BB_BASE_N + 1:i + 1]
            out.append((_mean(win), _std_pop(win)))
        else:
            out.append(None)
    return out


def _rsi_value(avg_gain: float, avg_loss: float) -> float:
    if avg_loss == 0:
        return 100.0 if avg_gain > 0 else 50.0
    if avg_gain == 0:
        return 0.0
    return 100.0 - 100.0 / (1.0 + avg_gain / avg_loss)


def _rsi_per_bar(closes, period: int = 14) -> list:
    """Wilder RSI per bar; None before the seed. Copy of
    rsi_stoch_reversion._rsi_per_bar (identical formula to marketstate's
    rsi14 over the local window)."""
    if len(closes) < period + 1:
        return [None] * len(closes)
    deltas = [closes[i] - closes[i - 1] for i in range(1, len(closes))]
    gains = [max(d, 0.0) for d in deltas]
    losses = [max(-d, 0.0) for d in deltas]
    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period
    out = [None] * period
    out.append(_rsi_value(avg_gain, avg_loss))
    for i in range(period, len(deltas)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period
        out.append(_rsi_value(avg_gain, avg_loss))
    return out


def _retracement_level(fibs: tuple, ratio: float):
    """The retracement level for `ratio` from the self-describing fib tuple
    (anchor, direction, retr, ext); None when absent (fib_retracement_
    continuation._retracement_level)."""
    if not isinstance(fibs, tuple) or len(fibs) != 4:
        return None
    for r, level in fibs[2]:
        if abs(r - ratio) < 1e-9:
            return float(level)
    return None


class FibRsiBbConfluenceExpert(Expert):
    """Three indicator families must agree (variant a) or a two-of-three
    majority must agree (variant b), or there is no setup."""
    expert_id = 'fib_rsi_bb_confluence'
    version = 'v1'
    mechanism_family_id = 'confluence_reversion_continuation'
    behavior_family_id = 'fib_rsi_bb_confluence'
    variant_id = 'a'
    requires = ('oscillator', 'location', 'volatility', 'history')
    # D-044: both confluence rules evaluated on the dev window, losers
    # included. D-046: every threshold is inherited from a registered family,
    # frozen pre-window — the search consumed exactly the two rules.
    variants_evaluated = ('a', 'b')
    search_universe_size = 2

    def __init__(self, variant_id: str | None = None):
        if variant_id is not None:
            if variant_id not in self.variants_evaluated:
                raise ValueError(
                    f'{self.expert_id}: unknown variant {variant_id!r} '
                    f'(variants_evaluated={list(self.variants_evaluated)})')
            self.variant_id = variant_id

    # --- per-leg votes -----------------------------------------------------
    # Each leg votes 'LONG' | 'SHORT' | None for a given bar, computed from
    # the local per-bar series (the anchor scan) or the state features (the
    # current bar). A leg that cannot form an opinion abstains (None).

    def _bb_vote_at(self, i: int, bar: tuple) -> str | None:
        """Fade-zone vote: close between 2-SD and 3-SD of the 20-bar SMA
        (bollinger_reversion Setup 2). A degenerate band (sd == 0) is no
        level; a close beyond 3-SD is a trend, not a reversion."""
        if i < BB_BASE_N - 1 or self._bb[i] is None:
            return None
        mid, sd = self._bb[i]
        if sd <= 0:
            return None
        close = bar[4]
        if mid - 3 * sd < close <= mid - 2 * sd:
            return 'LONG'
        if mid + 2 * sd <= close < mid + 3 * sd:
            return 'SHORT'
        return None

    def _rsi_vote_at(self, i: int) -> str | None:
        """Oscillator-reversion vote (rsi_stoch_reversion variant a): the
        newest run sits on the reverted side of its extreme AND the prior bar
        (the run start's predecessor) is on the extreme side — the dip that
        made the recovery meaningful. The close-beyond-signal-bar price
        confirmation of the pure oscillator strategy is subsumed here by the
        other legs' close conditions, so this leg carries no extra close
        constraint (documented deviation)."""
        if i >= len(self._rsi) or self._rsi[i] is None:
            return None
        # LONG checked first, mirroring rsi_stoch_reversion._detect_rsi's
        # return-early order; a mid-range bar that recovered from both sides
        # in the window resolves the same way as the source family.
        if self._rsi[i] > RSI_OS:
            s = i
            while s > 0 and self._rsi[s - 1] is not None \
                    and self._rsi[s - 1] > RSI_OS:
                s -= 1
            if s > 0 and self._rsi[s - 1] is not None \
                    and self._rsi[s - 1] <= RSI_OS:
                return 'LONG'
            # No oversold dip: fall through to the overbought side rather
            # than abstaining — a mid-range bar can be recovered from the
            # overbought extreme (mirrors rsi_stoch_reversion._detect_rsi,
            # whose LONG branch returns only on a dip and otherwise falls
            # through to the SHORT check).
        if self._rsi[i] < RSI_OB:
            s = i
            while s > 0 and self._rsi[s - 1] is not None \
                    and self._rsi[s - 1] < RSI_OB:
                s -= 1
            if s > 0 and self._rsi[s - 1] is not None \
                    and self._rsi[s - 1] >= RSI_OB:
                return 'SHORT'
        return None

    def _fib_vote_at(self, i: int, bar: tuple) -> str | None:
        """Deep-retracement reclaim vote (fib_retracement_continuation at
        0.786): a close that reclaimed the deepest retracement of the intact
        impulse. `_fib_level`/`_fib_direction` are frozen from the state's
        fib_levels at detection, so the anchor scan uses the same level."""
        if self._fib_direction is None or self._fib_level is None:
            return None
        _e, _o, high, low, close, _f, _s = bar
        if self._fib_direction == 1.0:
            return 'LONG' if (close > self._fib_level
                              and low <= self._fib_level) else None
        return 'SHORT' if (close < self._fib_level
                           and high >= self._fib_level) else None

    def _confluence_vote(self, votes: list) -> str | None:
        """The confluence rule for this variant over the three leg votes.
        variant a (STRICT): all three vote the same direction. variant b
        (MAJORITY): at least two of the three agree."""
        if self.variant_id == 'a':
            if all(v is not None for v in votes) \
                    and votes[0] == votes[1] == votes[2]:
                return votes[0]
            return None
        longs = sum(1 for v in votes if v == 'LONG')
        shorts = sum(1 for v in votes if v == 'SHORT')
        if longs >= 2:
            return 'LONG'
        if shorts >= 2:
            return 'SHORT'
        return None

    def _confluence_at(self, i: int, bar: tuple) -> str | None:
        """The per-bar confluence predicate for the anchor scan (D-026):
        the same variant rule over the three local-series votes."""
        return self._confluence_vote([self._bb_vote_at(i, bar),
                                      self._rsi_vote_at(i),
                                      self._fib_vote_at(i, bar)])

    # --- evaluate ----------------------------------------------------------

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.history',
                f'{sym}.bb_mid', f'{sym}.bb_upper', f'{sym}.bb_lower',
                f'{sym}.rsi14', f'{sym}.fib_levels']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        hist_value = f[f'{sym}.history'].value
        fibs = f[f'{sym}.fib_levels'].value
        if atr is None or not isinstance(hist_value, (tuple, list)) \
                or not hist_value or not isinstance(fibs, tuple) \
                or len(fibs) != 4:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        # Fib warmup needs a confirmed swing pair (n_close >= 21 in
        # marketstate); BB needs 20, RSI a seed. A shorter window cannot host
        # the confluence, so it is habitat-unavailable, not signal-less.
        if len(hist_value) < 21:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        anchor_price, direction, _retr, _ext = fibs
        if direction not in (1.0, -1.0) or anchor_price <= 0.0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        self._fib_level = _retracement_level(fibs, FIB_RATIO)
        if self._fib_level is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        self._fib_direction = direction
        self._hist = tuple(hist_value)
        self._bb = _bb_series(self._hist)
        self._rsi = _rsi_per_bar([b[4] for b in self._hist])

        # Current-bar votes. The RSI leg additionally requires the STATE
        # feature to clear the threshold (the full-series Wilder value can
        # disagree with the local window near the boundary; the conservative
        # gate from rsi_stoch_reversion).
        bb_v = self._bb_vote_at(len(self._hist) - 1, self._hist[-1])
        rsi_v = self._rsi_vote_at(len(self._hist) - 1)
        rsi_feat = float(f[f'{sym}.rsi14'].value)
        if rsi_v == 'LONG' and not rsi_feat > RSI_OS:
            rsi_v = None
        if rsi_v == 'SHORT' and not rsi_feat < RSI_OB:
            rsi_v = None
        fib_v = self._fib_vote_at(len(self._hist) - 1, self._hist[-1])
        direction_sig = self._confluence_vote([bb_v, rsi_v, fib_v])
        if direction_sig is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        anchor = self.find_setup_anchor(self._hist, self._confluence_at)

        # Geometry, frozen at detection (D-042 prior_*_ref pattern). The
        # 78.6% level doubles as the pre-entry invalidation and the
        # post-entry deep-correction reference; the frozen 3-SD band is the
        # reversion-premise reference. `variant` separates the episode keys.
        mid = float(f[f'{sym}.bb_mid'].value)
        upper = float(f[f'{sym}.bb_upper'].value)
        lower = float(f[f'{sym}.bb_lower'].value)
        geometry = self.declared_geometry()
        geometry.update({'atr_ref': atr, 'variant': self.variant_id})
        if direction_sig == 'LONG':
            geometry['prior_low_ref'] = self._fib_level
            geometry['lower_3sd_ref'] = mid - 1.5 * (mid - lower)
        else:
            geometry['prior_high_ref'] = self._fib_level
            geometry['upper_3sd_ref'] = mid + 1.5 * (upper - mid)
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction_sig,
            setup_fingerprint=f'{sym}:{self.variant_id}:{direction_sig}:'
                              f'{self._fib_level:.6f}:{close:.6f}',
            risk_geometry=geometry, birth_time=t,
            setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    # --- still_valid ---------------------------------------------------------

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The confluence thesis is "the reversion-reversal holds at the deep
        support". It dies when ANY of the three premises breaks: a close
        beyond the frozen 78.6% level (deep correction — the impulse is
        gone), a close beyond the frozen 3-SD band (a trend, not a
        reversion), or the oscillator re-entering its extreme zone (the
        reversion failed). Unobservable inputs fail open — an unreadable
        thesis is not a dead one."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        c = float(close.value) if (close is not None
                                   and close.value is not None) else None
        long = draft.direction == 'LONG'
        if c is not None:
            ref = draft.risk_geometry.get(
                'prior_low_ref' if long else 'prior_high_ref')
            if ref is not None:
                if (long and c < float(ref)) or (not long and c > float(ref)):
                    return False
            ref3 = draft.risk_geometry.get(
                'lower_3sd_ref' if long else 'upper_3sd_ref')
            if ref3 is not None:
                if (long and c <= float(ref3)) or (not long and c >= float(ref3)):
                    return False
        rsi = f.get(f'{sym}.rsi14')
        if rsi is not None and rsi.value is not None:
            v = float(rsi.value)
            if (long and v <= RSI_OS) or (not long and v >= RSI_OB):
                return False
        return True
